plc_ioc_status.py

```diff
@@ -0,0 +1,182 @@
+            plc_status_indicator.setColor(self._off_color)
+
+            # total initial number of ffs to initialize the dictionaries with
+            # num_ffo * num_ff
+            all_ffos = ((ffo_end - ffo_start) + 1) * (ff_end - ff_start + 1)
+            self.ffs_count_map[plc_name] = {'online': [False]*all_ffos,
+                                            'in_use': [False]*all_ffos,
+                                            'alarmed': [False]*all_ffos,
+                                            'plc_status': False}
+            self.ffs_label_map[plc_name] = {'online': label_online,
+                                            'in_use': label_in_use,
+                                            'alarmed': label_alarmed,
+                                            'plc_status': plc_status_indicator}
+
+            count = 0
+            for _ffo, _ff in entries:
+                s_ffo = str(_ffo).zfill(ffos_zfill)
+                s_ff = str(_ff).zfill(ffs_zfill)
+                ch_macros = dict(index=count, P=prefix, FFO=s_ffo, FF=s_ff)
+
+                ch = Template(
+                    'ca://${P}FFO:${FFO}:FF:${FF}:Info:InUse_RBV').safe_substitute(**ch_macros)
+                channel = PyDMChannel(
+                    ch,
+                    connection_slot=functools.partial(
+                        self.ffo_connection_callback, plc_name, count),
+                    value_slot=functools.partial(
+                        self.ffo_value_changed, plc_name, count),
+                    severity_slot=functools.partial(
+                        self.ffo_severity_changed, plc_name, count))
+                # should not be adding a new connection because this address
+                # already exists in the connections,
+                # instead should just add a listener
+                channel.connect()
+                count += 1
+
+            widget = QtWidgets.QWidget()
+            widget_layout = QtWidgets.QHBoxLayout()
+
+            widget.setLayout(widget_layout)
+            widget.layout().addWidget(label_name)
+            widget.layout().addWidget(label_online)
+            widget.layout().addWidget(label_in_use)
+            widget.layout().addWidget(label_alarmed)
+            widget.layout().addWidget(label_heartbeat)
+            widget.layout().addWidget(plc_status_indicator)
+
+            self.plc_ioc_container.layout().addWidget(widget)
+        vertical_spacer = (
+            QtWidgets.QSpacerItem(20, 40,
+                                  QtWidgets.QSizePolicy.Preferred,
+                                  QtWidgets.QSizePolicy.MinimumExpanding))
+        self.plc_ioc_container.layout().addItem(vertical_spacer)
+
+    def heartbeat_severity_changed(self, key, alarm):
+        # 0 = NO_ALARM, 1 = MINOR, 2 = MAJOR, 3 = INVALID, # 4 = DISCONNECTED
+        plc = self.ffs_count_map.get(key)
+        if alarm == 3:
+            plc['plc_status'] = False
+        else:
+            plc['plc_status'] = True
```

Comment:
I'm not sure if alarm status can ever be 4 here, but in that case I think plc_status should also show False (because in this case we do not know if the PLC is off or on)

Reply:
Yeah I need to investigate this, I thought PyDM will send a `ALARM_DISCONNECTED = 4` when the PV is disconnected but I might be wrong here, i am checking now - and will get back to this

Reply:
Ok just clarified with Hugo, the PyDM Widgets would have the `ALARM_DISCONNECTED` as well, but the `PyDMChannel` does not - so I am good here with not checking for 4, but I should adjust the comment above. 

Thank you

Reply:
ahh okay that makes sense

Reply:
The `ALARM_DISCONNECTED` is an artificial flag used for PyDM widgets based on the `connection` information from the channel. EPICS itself has 0-3 only.

Reply:
Thanks Hugo ;)